### PR TITLE
[controller][server] Move ideal state listener to EV repository and read Store Metadata for CV updates

### PR DESCRIPTION
**Pull request by majisourav99**
## Move ideal state listener to EV repository and read Store Metadata for CV updates

ExternalView(EV) and CustomizedView (CV) repository listens to IdealState from Helix to filter invalid state changes. This add a lots of delay to execute those callbacks (Helix callbackHandler).  Currently EV is used by the controller and CV is used by router and servers to query replica statuses and mostly server and routers would be exposed to the problem of slow transitions. 

This PR
1.  moves the ideal state watch to EV.
2.  make CV repository filter invalid states by query store version metadata. 
That way updates router and servers host wont add many callbacks to helix and hopefully results in faster state transitions.

## How was this PR tested?

test WIP
## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixExternalViewRepository.java [hunk 1]
```diff
@@ -32,9 +34,11 @@
  * Extend HelixBaseRoutingRepository to leverage external view data.
  */
 @BatchMode
-public class HelixExternalViewRepository extends HelixBaseRoutingRepository {
+public class HelixExternalViewRepository extends HelixBaseRoutingRepository implements IdealStateChangeListener {
   private static final Logger LOGGER = LogManager.getLogger(HelixExternalViewRepository.class);
 
+  protected volatile Map<String, Integer> resourceToIdealPartitionCountMap;
```
Comment (lluwm) [inline internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixExternalViewRepository.java:40]: I know it is the existing issue, but if possible, I would prefer to declare it as `AtomicReference`, as volatile only solves the visibility issue, thus making the change to `resourceToIdealPartitionCountMap` visible to all threads, however it doesn't guarantee an atomic update.

**Review by lluwm — COMMENTED** [review]
Thank you and the rb looks good to me! Do you have a plan to run some Frankenwar tests to verify that the number of Idealstate callbacks are reduced after this change?

File: internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixExternalViewRepository.java (hunk 1 above)
Comment (majisourav99) [reply internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixExternalViewRepository.java:40]: good idea, let me update it.

**Comment by sixpluszero** [comment]
Can you add some more unit tests? UnitTest Coverage seems failing for now.

File: internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java [hunk 2]
```diff
@@ -107,44 +113,31 @@ protected void onCustomizedViewDataChange(RoutingTableSnapshot routingTableSnaps
...
+          if (store == null) {
+            LOGGER.warn("Cannot find store for resource: {}.", resourceName);
+            continue;
+          }
+          boolean isValidVersion = store.getVersions().stream().anyMatch(v -> v.getNumber() == version);
+          if (!isValidVersion) {
+            LOGGER.warn("Version not found in store for resource: {}.", resourceName);
+            continue;
+          }
+          partitionCount = store.getVersion(version).get().getPartitionCount();
```
Comment (FelixGV) [inline internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java:136]: Is it necessary to do this stream? Wouldn't it be simpler to do:

```
  Optional<Version> optionalVersion = store.getVersion(version);
  if (!optionalVersion.isPresent()) {
    // log/continue
  }
  partitionCount = optionalVersion.get().getPartitionCount();
```

Same number of lines but anyway... as you wish.

File: internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java [hunk 3]
```diff
@@ -107,44 +113,31 @@ protected void onCustomizedViewDataChange(RoutingTableSnapshot routingTableSnaps
...
+          resourceToPartitionCountMap.put(resourceName, partitionCount);
```
Comment (FelixGV) [inline internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java:137]: I see puts on this final map but never any remove or clear. Is it a mem leak? Should we instead make the map non-final, build a new one in this function here, and swap the whole map?

File: internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixExternalViewRepository.java (hunk 1 above)
Comment (FelixGV) [reply internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixExternalViewRepository.java:40]: Are you saying we should update this variable with CAS? Is there a need for that?

If we don't do CAS, and only use the `get` and `set` functions on `AtomicReference`, then it is literally just setting and getting a volatile field inside the `AR`.

File: internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/TestBatch.java [hunk 4]
```diff
@@ -231,9 +231,10 @@ public void testDataPushWithSchemaWithAWrongDefault(boolean extendedSchemaValida
     }
   }
 
-  @Test(dataProvider = "Two-True-and-False", dataProviderClass = DataProviderUtils.class)
-  public void testCompressingRecord(boolean compressionMetricCollectionEnabled, boolean useMapperToBuildDict)
-      throws Exception {
+  @Test
+  public void testCompressingRecord() throws Exception {
+    boolean compressionMetricCollectionEnabled = false;
+    boolean useMapperToBuildDict = false;
```
Comment (FelixGV) [inline internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/TestBatch.java:237]: Why this change?

**Review by FelixGV — COMMENTED** [review]
Looks good overall... just some minor comments.

File: internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/TestBatch.java (hunk 4 above)
Comment (majisourav99) [reply internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/TestBatch.java:237]: unintended, will remove.

File: internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java (hunk 2 above)
Comment (majisourav99) [reply internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java:136]: BTW I had a concern regarding race condition in getting store metadata up vs `onCustomizedViewDataChange` call. Could we miss update due to delay in store metadata?

Comment (FelixGV) [reply internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java:136]: Maybe...!

File: internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixExternalViewRepository.java (hunk 1 above)
Comment (majisourav99) [reply internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixExternalViewRepository.java:40]: Right, I discussed offline with Lei, there are no differences to use AR vs `volatile` in this case.

File: internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java [hunk 5]
```diff
@@ -52,7 +51,44 @@ public HelixCustomizedViewOfflinePushRepository(SafeHelixManager manager, ReadOn
     super(manager);
     dataSource.put(PropertyType.CUSTOMIZEDVIEW, Collections.singletonList(HelixPartitionState.OFFLINE_PUSH.name()));
     this.storeRepository = storeRepository;
-    this.storeRepository.registerStoreDataChangedListener(this);
+    this.storeRepository.registerStoreDataChangedListener(new StoreDataChangedListener() {
```
Comment (FelixGV) [inline internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java:54]: Hmm... I guess this is just a nitpick, but why move the code from normal class functions to an anonymous class here? The previous form looked a bit cleaner IMHO. Anonymous classes make stacktraces a bit harder to interpret... is there a benefit I'm not thinking of?

File: internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java [hunk 6]
```diff
@@ -40,9 +43,52 @@ public class HelixCustomizedViewOfflinePushRepository extends HelixBaseRoutingRe
...
+      public void handleStoreChanged(Store store) {
+        int currentVersion = store.getCurrentVersion();
+        if (currentVersion == NON_EXISTING_VERSION) {
+          return;
+        }
+
+        resourceToPartitionCountMap.entrySet().removeIf(entry -> {
+          String storeName = Version.parseStoreFromKafkaTopicName(entry.getKey());
+          int version = Version.parseVersionFromVersionTopicName(entry.getKey());
+          return store.getName().equals(storeName) && version != currentVersion;
```
Comment (FelixGV) [inline internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java:77]: Is it correct to keep only the current version? Do we not care for the other existing (but non-current) versions? I am not sure...

Comment (majisourav99) [reply internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java:77]: For CV, does future version matter? backup most definitely not.  To be safe I will just delete non-existent versions.

File: internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java (hunk 5 above)
Comment (majisourav99) [reply internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java]: Converted it to inner class as discussed offline.

**Review by FelixGV — APPROVED** [review]
Thanks!

File: internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java [hunk 7]
```diff
@@ -221,7 +221,12 @@ public void refreshRoutingDataForResource(String kafkaTopic) {
     throw new VeniceException("The function of refreshRoutingDataForResource is not implemented");
   }
 
-  private class StoreChangeListener implements StoreDataChangedListener {
+  // test only
+  public Map<String, Integer> getResourceToPartitionCountMap() {
```
Comment (FelixGV) [inline internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java:225]: Can we make this package private if it's for tests only? Also, just to be extra cautious regarding misuse, it may be good to wrap it in `Collections:unmodifiableMap`.

File: internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixReadOnlyStoreRepositoryAdapter.java [hunk 8]
```diff
@@ -84,6 +84,11 @@ public Store getStoreOrThrow(String storeName) throws VeniceNoStoreException {
     return store;
   }
 
+  // test only
+  public Set<StoreDataChangedListener> getListeners() {
```
Comment (FelixGV) [inline internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixReadOnlyStoreRepositoryAdapter.java:88]: Same here? Can we make it package private and use `Collections::unmodifiableSet`?

**Review by FelixGV — COMMENTED** [review]
Thanks, LGTM, just a couple of minor comments. Please take a look, I think we can merge this soon.

File: internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java (hunk 7 above)
Comment (majisourav99) [reply internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java:225]: Sure, good idea to wrap it unmodifiable!

**Review by FelixGV — APPROVED** [review]
Thanks Sourav!
